In [36]:
import pandas as pd

In [37]:
file_name = "titanic_data.csv"
pd.read_csv(file_name)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [38]:
df = pd.read_csv(file_name)


In [39]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [41]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [42]:
missing = df.isnull().sum()
columns_to_clean = missing[missing > 0]
columns_to_clean

Age         177
Cabin       687
Embarked      2
dtype: int64

In [43]:
df.drop(columns=['Cabin'], inplace=True)

In [44]:
df['Age'] = df['Age'].fillna(df['Age'].median())

In [45]:
df['Embarked'] = df['Embarked'].fillna(df['Embarked'].mode()[0])

In [46]:
df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

In [47]:
rows_before = df.shape[0]
rows_before

891

In [48]:
df.drop_duplicates(inplace=True)

In [49]:
rows_after = df.shape[0]
rows_after

891

In [50]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Embarked        object
dtype: object

In [51]:
df['Survived'] = df['Survived'].astype('category')
df['Pclass'] = df['Pclass'].astype('category')
df['Sex'] = df['Sex'].astype('category')
df['Embarked'] = df['Embarked'].astype('category')

In [52]:
df.dtypes

PassengerId       int64
Survived       category
Pclass         category
Name             object
Sex            category
Age             float64
SibSp             int64
Parch             int64
Ticket           object
Fare            float64
Embarked       category
dtype: object

In [53]:
df['Fare_Band'] = pd.cut(
    df['Fare'],
    bins=[0, 10, 30, 100, 600],
    labels=['Low', 'Medium', 'High', 'Very High']
)

df[['Fare', 'Fare_Band']].head()

Fare Fare_Band
0   7.2500       Low
1  71.2833      High
2   7.9250       Low
3  53.1000      High
4   8.0500       Low

In [54]:
def age_category(age):
    if age < 18:
        return "Child"
    elif age < 60:
        return "Adult"
    else:
        return "Senior"

df['Age_Category'] = df['Age'].apply(age_category)

df[['Age', 'Age_Category']].head()

Age Age_Category
0  22.0        Adult
1  38.0        Adult
2  26.0        Adult
3  35.0        Adult
4  35.0        Adult

In [55]:
df.head()

PassengerId Survived Pclass  \
0            1        0      3   
1            2        1      1   
2            3        1      3   
3            4        1      1   
4            5        0      3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked Fare_Band Age_Category  
0      0         A/5 21171   7.2500        S       Low        Adult  
1      0          PC 17599  71.2833        C      High        Adult  
2      0  STON/O2. 3101282   7.9250        S       Low        Adult  
3      0            113803  53.1000        S      High        Adult  
4      0            373450   8.0500        S       Low        Adult

**Titanic Dataset:**
Data Cleaning Summary

**1. Identifying Missing Values**

* Missing values were checked using `.isnull().sum()`.
* The Age column had 177 missing values, Embarked had 2 missing values, and Cabin had 687 missing values.
* Columns with missing values were analyzed to decide whether to fill or drop them before modeling or reporting.

**2. Handling Missing Values**

* Age missing values were filled using the **median** to reduce the effect of outliers.
* Embarked missing values were filled using the **mode** since it is categorical.
* Cabin was **dropped** due to too many missing entries.

**3. Removing Duplicates**

* Duplicate rows were checked and removed to ensure the dataset remained accurate and clean.

**4. Correcting Data Types**

* Columns like Survived, Pclass, Sex, and Embarked were converted to **category** type for better analysis.
* No date columns were present, so datetime conversion was not needed.

**5. Feature Engineering**

* A new column, **Age_Category**, was created to group passengers into Child, Adult, and Senior.
* Another column, **Fare_Band**, was created to classify ticket prices as Low, Medium, High, or Very High.